src/Godot/Api/Types.hs

```diff
@@ -0,0 +1,6642 @@
+        EditorInspectorPlugin(..), EditorInspector(..),
+        AnimationTrackEditPlugin(..), JavaScript(..),
+        ScriptCreateDialog(..), CSGShape(..), EditorScenePostImport(..),
+        CSGCylinder(..), CSGSphere(..), CSGMesh(..), CSGPrimitive(..),
+        CSGBox(..), CSGPolygon(..), CSGTorus(..),
+        NetworkedMultiplayerENet(..), CSGCombiner(..),
+        ResourceFormatDDS(..), GDNativeLibraryResourceLoader(..),
+        GDNativeLibraryResourceSaver(..), MultiplayerPeerGDNative(..),
+        PacketPeerGDNative(..), StreamPeerGDNative(..),
+        ARVRInterfaceGDNative(..), GDNative(..), ResourceFormatPKM(..),
+        GDNativeLibrary(..), ResourceFormatSaverNativeScript(..),
+        ResourceFormatLoaderVideoStreamGDNative(..),
+        GDScriptFunctionState(..), PluginScript(..), NativeScript(..),
+        ResourceFormatLoaderNativeScript(..), GDScript(..),
+        VideoStreamGDNative(..), ResourceFormatSaverGDScript(..),
+        GridMap(..), ResourceFormatLoaderGDScript(..),
+        MobileVRInterface(..), OpenSimplexNoise(..), RegEx(..),
+        ResourceImporterOGGVorbis(..), ResourceImporter(..),
+        AudioStreamOGGVorbis(..), NoiseTexture(..), ResourceFormatPVR(..),
+        RegExMatch(..), VideoStreamTheora(..),
+        ResourceFormatLoaderTheora(..), UPNP(..), UPNPDevice(..),
+        VisualScriptNode(..), VisualScriptFunctionState(..),
+        VisualScriptFunction(..), VisualScript(..),
+        VisualScriptVariableGet(..), VisualScriptOperator(..),
+        VisualScriptClassConstant(..), VisualScriptIndexGet(..),
+        VisualScriptVariableSet(..), VisualScriptMathConstant(..),
+        VisualScriptIndexSet(..), VisualScriptGlobalConstant(..),
+        VisualScriptConstant(..), VisualScriptSceneNode(..),
+        VisualScriptEngineSingleton(..), VisualScriptLocalVar(..),
+        VisualScriptSelf(..), VisualScriptResourcePath(..),
+        VisualScriptBasicTypeConstant(..), VisualScriptSceneTree(..),
+        VisualScriptSubCall(..), VisualScriptComment(..),
+        VisualScriptConstructor(..), VisualScriptCustomNode(..),
+        VisualScriptInputAction(..), VisualScriptLocalVarSet(..),
+        VisualScriptFunctionCall(..), VisualScriptTypeCast(..),
+        VisualScriptDeconstruct(..), VisualScriptPreload(..),
+        VisualScriptPropertySet(..), VisualScriptWhile(..),
+        VisualScriptIterator(..), VisualScriptSwitch(..),
+        VisualScriptSelect(..), VisualScriptSequence(..),
+        VisualScriptPropertyGet(..), VisualScriptEmitSignal(..),
+        VisualScriptReturn(..), VisualScriptCondition(..),
+        VisualScriptYieldSignal(..), VisualScriptBuiltinFunc(..),
+        VisualScriptYield(..), WebSocketServer(..),
+        Physics2DDirectBodyStateSW(..), BulletPhysicsDirectBodyState(..),
+        WebSocketMultiplayerPeer(..), VideoStreamWebm(..),
+        ResourceFormatLoaderWebm(..), BulletPhysicsServer(..),
+        WebSocketClient(..), WebSocketPeer(..), Physics2DServerSW(..))
```

Comment:
I took the liberty to fix this in 56f4e80272503822e72085ef1e9b154e7f62be9b

Reply:
Demo needs to be updated after this.

Reply:
As a user, I would greatly prefer that all types be available.
There are so many packages that are totally broken because there's some advanced functionality that one needs but the type constructors aren't exposed.

It's your call. But I personally think the library is totally broken and unusable if you hide type constructors.

Reply:
Casting to a parent type is very easy with `upcast`, it's only when you want to cast something to a type you don't know for sure is correct that is impacted. Since it may fail at runtime it should be represented as such at the type level via `tryCast`. You may argue that the slight impact on performance isn't worth it, in which case a compromise could be made by exposing the constructors from a module with a discouraging namespace at least.

Reply:
Do you have any use cases in mind where you can't do without type constructors (or where it's a significant inconvenience)? I'd love to see examples so that I can form a more educated opinion.